broadcaster/backlog/backlog.go

```diff
@@ -0,0 +1,399 @@
+	return &backlogSegment{
+		messages: []*m.BroadcastFeedMessage{},
+	}
+}
+
+// IsBacklogSegmentNil uses the internal backlogSegment type to check if a
+// variable of type BacklogSegment is nil or not. Comparing whether an
+// interface is nil directly will not work.
+func IsBacklogSegmentNil(segment BacklogSegment) bool {
+	return segment.(*backlogSegment) == nil
+}
+
+// Start returns the first message index within the backlogSegment.
+func (s *backlogSegment) Start() uint64 {
+	return uint64(s.start.Load())
+}
+
+// End returns the last message index within the backlogSegment.
+func (s *backlogSegment) End() uint64 {
+	return uint64(s.end.Load())
+}
+
+// Next returns the next backlogSegment.
+func (s *backlogSegment) Next() BacklogSegment {
+	return s.nextSegment.Load()
+}
+
+// Messages returns all of the messages stored in the backlogSegment.
+func (s *backlogSegment) Messages() []*m.BroadcastFeedMessage {
+	s.messagesLock.RLock()
+	defer s.messagesLock.RUnlock()
+	return s.messages
+}
+
+// get reads messages from the given start to end message index.
+func (s *backlogSegment) get(start, end uint64) ([]*m.BroadcastFeedMessage, error) {
+	noMsgs := []*m.BroadcastFeedMessage{}
+	if start < s.start.Load() {
+		return noMsgs, errOutOfBounds
+	}
+
+	if end > s.end.Load() {
+		return noMsgs, errOutOfBounds
+	}
+
+	startIndex := start - s.start.Load()
+	endIndex := end - s.start.Load() + 1
+
+	s.messagesLock.RLock()
+	defer s.messagesLock.RUnlock()
+	return s.messages[startIndex:endIndex], nil
+}
+
+// append appends the given BroadcastFeedMessage to messages if it is the first
+// message in the sequence or the next in the sequence. If segment's end
+// message is ahead of the given message append will do nothing. If the given
+// message is ahead of the segment's end message append will return
+// errDropSegments to ensure any messages before the given message are dropped.
+func (s *backlogSegment) append(prevMsgIdx uint64, msg *m.BroadcastFeedMessage) error {
+	s.messagesLock.Lock()
+	defer s.messagesLock.Unlock()
+	seen := false
+	defer s.updateSegment(&seen)
+
+	if expSeqNum := prevMsgIdx + 1; prevMsgIdx == 0 || uint64(msg.SequenceNumber) == expSeqNum {
+		s.messages = append(s.messages, msg)
+	} else if uint64(msg.SequenceNumber) > expSeqNum {
+		s.messages = nil
+		s.messages = append(s.messages, msg)
+		return fmt.Errorf("new message sequence number (%d) is greater than the expected sequence number (%d): %w", msg.SequenceNumber, expSeqNum, errDropSegments)
+	} else {
+		seen = true
+		return errSequenceNumberSeen
+	}
+	return nil
+}
+
+// Contains confirms whether the segment contains a message with the given
+// sequence number.
+func (s *backlogSegment) Contains(i uint64) bool {
+	start := s.start.Load()
+	if i < start || i > s.end.Load() {
+		return false
+	}
+
+	msgIndex := i - start
+	s.messagesLock.RLock()
+	msg := s.messages[msgIndex]
+	s.messagesLock.RUnlock()
+	return uint64(msg.SequenceNumber) == i
```

Comment:
There's the same race condition here as in `get` of the messages potentially being cleared between the start/end check and actually getting the message. Personally I'd probably only check start/end in this function and not check the messages array, but either way works.

Reply:
Thanks I have moved the lock to cover the whole function, I believe that should fix the race condition but let me know if you think I should do a different implementation :)